### PR TITLE
Add unit test coverage requirements for some packages

### PR DESCRIPTION
**Pull request by lunny**
This PR add a CI check for package tests coverage. When the coverage is less than a previous value, CI will be failure.

### CODE REVIEW

**Comment by KN4CK3R** [comment]
I think this should use the merged results. Currently it only uses the integration test result.

**Comment by lunny** [comment]
> I think this should use the merged results. Currently it only uses the integration test result.

Done

File: Makefile [hunk 1]
```diff
@@ -394,13 +395,19 @@ test\#%:
...
 .PHONY: unit-test-coverage
 unit-test-coverage:
 	@echo "Running unit-test-coverage $(GOTESTFLAGS) -tags '$(TEST_TAGS)'..."
 	@$(GO) test $(GOTESTFLAGS) -timeout=20m -tags='$(TEST_TAGS)' -cover -coverprofile coverage.out $(GO_TEST_PACKAGES) && echo "\n==>\033[32m Ok\033[m\n" || exit 1
 
+.PHONY: test-coverage-check
+test-coverage-check:
+	@$(GO) run build/gocoverage.go check coverage.all \
+		code.gitea.io/gitea/modules/setting=7.1% \
+		code.gitea.io/gitea/modules/hostmatcher=68.6%
```
Comment (KN4CK3R) [inline Makefile]: I don't know if it is a good idea to "store" the results in the Makefile.
Anyway here are some more entries for the current code base:
```
code.gitea.io/gitea/modules/activitypub=60.66%
code.gitea.io/gitea/modules/analyze=73.91%
code.gitea.io/gitea/modules/auth=60.29%
code.gitea.io/gitea/modules/avatar=55.20%
code.gitea.io/gitea/modules/base=92.96%
code.gitea.io/gitea/modules/cache=14.60%
code.gitea.io/gitea/modules/charset=56.67%
code.gitea.io/gitea/modules/container=100.00%
code.gitea.io/gitea/modules/context=58.91%
code.gitea.io/gitea/modules/csv=100.00%
code.gitea.io/gitea/modules/doctor=20.87%
code.gitea.io/gitea/modules/emoji=85.11%
code.gitea.io/gitea/modules/eventsource=53.88%
code.gitea.io/gitea/modules/generate=0.00%
code.gitea.io/gitea/modules/git=61.93%
code.gitea.io/gitea/modules/gitgraph=75.77%
code.gitea.io/gitea/modules/graceful=29.29%
code.gitea.io/gitea/modules/hcaptcha=0.00%
code.gitea.io/gitea/modules/highlight=67.14%
code.gitea.io/gitea/modules/hostmatcher=79.34%
code.gitea.io/gitea/modules/html=100.00%
code.gitea.io/gitea/modules/httpcache=78.95%
code.gitea.io/gitea/modules/httplib=67.86%
code.gitea.io/gitea/modules/indexer=39.21%
code.gitea.io/gitea/modules/issue=75.47%
code.gitea.io/gitea/modules/json=43.75%
code.gitea.io/gitea/modules/lfs=70.98%
code.gitea.io/gitea/modules/log=69.41%
code.gitea.io/gitea/modules/markup=53.32%
code.gitea.io/gitea/modules/mcaptcha=0.00%
code.gitea.io/gitea/modules/metrics=0.00%
code.gitea.io/gitea/modules/migration=70.45%
code.gitea.io/gitea/modules/mirror=16.67%
code.gitea.io/gitea/modules/nosql=29.21%
code.gitea.io/gitea/modules/notification=64.94%
code.gitea.io/gitea/modules/options=48.74%
code.gitea.io/gitea/modules/packages=70.63%
code.gitea.io/gitea/modules/paginator=100.00%
code.gitea.io/gitea/modules/password=66.18%
code.gitea.io/gitea/modules/pprof=0.00%
code.gitea.io/gitea/modules/private=13.40%
code.gitea.io/gitea/modules/process=40.28%
code.gitea.io/gitea/modules/proxy=21.43%
code.gitea.io/gitea/modules/proxyprotocol=0.00%
code.gitea.io/gitea/modules/public=53.26%
code.gitea.io/gitea/modules/queue=39.17%
code.gitea.io/gitea/modules/recaptcha=0.00%
code.gitea.io/gitea/modules/references=80.50%
code.gitea.io/gitea/modules/regexplru=65.22%
code.gitea.io/gitea/modules/repository=49.68%
code.gitea.io/gitea/modules/secret=64.00%
code.gitea.io/gitea/modules/session=18.63%
code.gitea.io/gitea/modules/setting=54.19%
code.gitea.io/gitea/modules/sitemap=82.35%
code.gitea.io/gitea/modules/ssh=45.98%
code.gitea.io/gitea/modules/storage=57.14%
code.gitea.io/gitea/modules/structs=54.22%
code.gitea.io/gitea/modules/svg=96.15%
code.gitea.io/gitea/modules/sync=100.00%
code.gitea.io/gitea/modules/system=70.00%
code.gitea.io/gitea/modules/templates=43.01%
code.gitea.io/gitea/modules/test=68.83%
code.gitea.io/gitea/modules/timeutil=80.80%
code.gitea.io/gitea/modules/translation=60.70%
code.gitea.io/gitea/modules/typesniffer=92.86%
code.gitea.io/gitea/modules/updatechecker=0.00%
code.gitea.io/gitea/modules/upload=73.85%
code.gitea.io/gitea/modules/uri=27.78%
code.gitea.io/gitea/modules/user=23.53%
code.gitea.io/gitea/modules/util=72.93%
code.gitea.io/gitea/modules/validation=80.12%
code.gitea.io/gitea/modules/watcher=0.00%
code.gitea.io/gitea/modules/web=76.76%
code.gitea.io/gitea/modules/webhook=13.79%
code.gitea.io/gitea/models/activities=58.48%
code.gitea.io/gitea/models/admin=33.55%
code.gitea.io/gitea/models/asymkey=41.40%
code.gitea.io/gitea/models/auth=46.36%
code.gitea.io/gitea/models/avatars=31.82%
code.gitea.io/gitea/models/db=41.53%
code.gitea.io/gitea/models/git=44.19%
code.gitea.io/gitea/models/issues=57.31%
code.gitea.io/gitea/models/organization=71.00%
code.gitea.io/gitea/models/packages=71.66%
code.gitea.io/gitea/models/perm=57.05%
code.gitea.io/gitea/models/project=31.85%
code.gitea.io/gitea/models/pull=36.13%
code.gitea.io/gitea/models/repo=58.62%
code.gitea.io/gitea/models/secret=24.56%
code.gitea.io/gitea/models/system=56.16%
code.gitea.io/gitea/models/unit=48.72%
code.gitea.io/gitea/models/unittest=31.91%
code.gitea.io/gitea/models/user=55.82%
code.gitea.io/gitea/models/webhook=66.02%
code.gitea.io/gitea/routers/api=60.66%
code.gitea.io/gitea/routers/common=61.15%
code.gitea.io/gitea/routers/install=7.39%
code.gitea.io/gitea/routers/private=34.74%
code.gitea.io/gitea/routers/utils=100.00%
code.gitea.io/gitea/routers/web=28.95%
code.gitea.io/gitea/services/agit=56.59%
code.gitea.io/gitea/services/asymkey=36.73%
code.gitea.io/gitea/services/attachment=60.87%
code.gitea.io/gitea/services/auth=38.99%
code.gitea.io/gitea/services/automerge=44.83%
code.gitea.io/gitea/services/context=47.22%
code.gitea.io/gitea/services/convert=74.80%
code.gitea.io/gitea/services/cron=61.44%
code.gitea.io/gitea/services/externalaccount=0.00%
code.gitea.io/gitea/services/forms=36.81%
code.gitea.io/gitea/services/gitdiff=70.52%
code.gitea.io/gitea/services/issue=40.50%
code.gitea.io/gitea/services/lfs=57.06%
code.gitea.io/gitea/services/mailer=54.55%
code.gitea.io/gitea/services/markup=100.00%
code.gitea.io/gitea/services/migrations=35.97%
code.gitea.io/gitea/services/mirror=29.83%
code.gitea.io/gitea/services/org=49.02%
code.gitea.io/gitea/services/packages=51.94%
code.gitea.io/gitea/services/pull=43.34%
code.gitea.io/gitea/services/release=45.09%
code.gitea.io/gitea/services/repository=41.19%
code.gitea.io/gitea/services/task=43.75%
code.gitea.io/gitea/services/user=46.24%
code.gitea.io/gitea/services/webhook=66.63%
code.gitea.io/gitea/services/wiki=56.41%
```

**Comment by KN4CK3R** [comment]
```
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/validation coverage is 79.4%, required 80.1%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/webhook coverage is 13.3%, required 13.7%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/svg coverage is 94.4%, required 96.1%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/doctor coverage is 15.1%, required 20.8%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/cache coverage is 14.5%, required 14.6%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/emoji coverage is 83.3%, required 85.1%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/auth coverage is NaN%, required 60.2%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/graceful coverage is 29.0%, required 29.2%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/indexer coverage is NaN%, required 39.2%
2023/01/16 09:02:54 package code.gitea.io/gitea/services/cron coverage is 56.2%, required 61.4%
2023/01/16 09:02:54 package code.gitea.io/gitea/services/forms coverage is 36.5%, required 36.8%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/proxy coverage is 17.2%, required 21.4%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/process coverage is 38.5%, required 40.2%
2023/01/16 09:02:54 package code.gitea.io/gitea/services/convert coverage is 73.5%, required 74.8%
2023/01/16 09:02:54 package code.gitea.io/gitea/modules/issue coverage is NaN%, required 75.4%
2023/01/16 09:02:54 package code.gitea.io/gitea/routers/api coverage is NaN%, required 60.6%
2023/01/16 09:02:54 package code.gitea.io/gitea/routers/install coverage is 5.7%, required 7.3%
```
Now see it works. But it looks like our code calculates the coverage different than CodeCov.

**Comment by KN4CK3R** [comment]
Looks like CodeCov combines the results of the subdirectories. Our calculation displays the result for the files in that directory only (?). That would explain the NaN% lines.

**Comment by 6543** [comment]
I would lower the current requirements to match what we have atm ... and then do increase them by case

**Review by 6543 — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -396,13 +397,143 @@ test\#%:
...
+		code.gitea.io/gitea/services/mirror=29.83% \
+		code.gitea.io/gitea/services/org=49.02% \
+		code.gitea.io/gitea/services/packages=51.94% \
+		code.gitea.io/gitea/services/pull=43.34% \
+		code.gitea.io/gitea/services/release=45.09% \
+		code.gitea.io/gitea/services/repository=41.19% \
+		code.gitea.io/gitea/services/task=43.75% \
+		code.gitea.io/gitea/services/user=46.24% \
+		code.gitea.io/gitea/services/webhook=66.63% \
+		code.gitea.io/gitea/services/wiki=56.41% \
```
Comment (silverwind) [inline Makefile]: It seems clumsy to have these statically in the Makefile. How about writing them out to a file, loading it from there and also provide a `make` target that runs a bash script to update them to current coverage results. Ideally that update target could run in scheduled job weekly.

**Comment by silverwind** [comment]
> When the coverage is less than a previous value, CI will be failure

I forsee some PR friction if we add this. I'd guess more PRs than usual would become abandoned with such strict requirements as not every bug fix is easily testable. We can try, but I guess there should be a mechanism for a owner to override this requirement for individual PRs.

**Comment by wxiaoguang** [comment]
Is it the right way to go?

**Comment by 6543** [comment]
I think in worst case we can lower the number for a fix ... but we should enforce it ... we want good tested code

**Comment by wxiaoguang** [comment]
I agree to improve test code (when I review, I frequently suggest to add tests ...)

But, is it the right way to go? Filling the Makefile with a lot of percentage numbers?

**Comment by 6543** [comment]
we could improve the tool that checks it and add a new .min_coverage.conf or something like that witch do contain the threashhold

**Comment by wxiaoguang** [comment]
A clear Makefile could L-G-T-M

File: .github/workflows/pull-db-tests.yml [hunk 3]
```diff
@@ -251,3 +263,16 @@ jobs:
         env:
           TAGS: bindata
           USE_REPO_TEST_DIR: 1
+
+  test-coverage-check:
+    if: needs.files-changed.outputs.backend == 'true' || needs.files-changed.outputs.frontend == 'true'
```
Comment (silverwind) [inline .github/workflows/pull-db-tests.yml]: I still think the frontend dependency is not right in these backend jobs, see https://github.com/go-gitea/gitea/pull/25150 because frontend (js,css) can not influence backend in any way.

Comment (lunny) [reply .github/workflows/pull-db-tests.yml]: They will affect web integration tests because some of these tests will parse output HTML

File: .drone.yml [hunk 4]
```diff
@@ -1,1612 +0,0 @@
----
-kind: pipeline
```
Comment (6543) [inline .drone.yml]: unrelated?

File: .github/workflows/pull-db-tests.yml (hunk 3 above)
Comment (silverwind) [reply .github/workflows/pull-db-tests.yml]: But `frontend` does not produce/alter any HTML responses.

Comment (lunny) [reply .github/workflows/pull-db-tests.yml]: But templates and vue will? They will change the dom nodes dynamically?

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml]: Yes, but these modifications are not done when the backend tests run as backend tests do not run any javascript, they just assert the static HTML returned by the backend.

Vue rendering or any other JS behaviour can only be tested in E2E tests or using Vitest.

Comment (lunny) [reply .github/workflows/pull-db-tests.yml]: > Yes, but these modifications are not done when the backend tests run as backend tests do not run any javascript.

Except e2e tests?

Comment (wxiaoguang) [inline .github/workflows/pull-db-tests.yml]: I do not think the `frontend` would affect backedn test coverage.

```
            frontend:
              - "**/*.js"
              - "web_src/**"
              - "package.json"
              - "package-lock.json"
```

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml]: See https://github.com/go-gitea/gitea/pull/22401#discussion_r1227931883 and https://github.com/go-gitea/gitea/pull/25150 for the frontend dependency removal.

Comment (wxiaoguang) [reply .github/workflows/pull-db-tests.yml]: Yup, that comment seems to have been marked as "resolved", so I started this new conversation.

File: build/gocoverage.go [hunk 5]
```diff
@@ -116,3 +159,88 @@ func main() {
...
 	dumpProfiles(merged, os.Stdout)
 }
+
+func percentToInt64(percent string) int64 {
+	value := strings.ReplaceAll(percent, "%", "")
+	i, err := strconv.ParseFloat(value, 10)
+	if err != nil {
+		log.Fatalf("invalid percent: %s", percent)
+	}
+	return int64(i * 10)
```
Comment (wxiaoguang) [inline build/gocoverage.go:169]: Why `i*10` ?  `12.34%` means "123"? That's a strange unit.

File: build/gocoverage.go [hunk 6]
```diff
@@ -116,3 +159,88 @@ func main() {
...
+		parts := strings.Split(p, "=")
+		if len(parts) != 2 {
+			continue
+		}
+		packagesRequirements[parts[0]] = percentToInt64(parts[1])
+	}
+	packagesTotals := make(map[string]int64)
+	packagesActives := make(map[string]int64)
+	for _, p := range profiles {
+		pkg := path.Dir(p.FileName)
```
Comment (wxiaoguang) [inline build/gocoverage.go]: filepath for file path

File: build/gocoverage.go [hunk 7]
```diff
@@ -116,3 +159,88 @@ func main() {
...
+		}
+		total, active := profileCount(p)
+		packagesTotals[pkg] += total
+		packagesActives[pkg] += active
+	}
+	var failed bool
+	for k, v := range packagesRequirements {
+		actual := 100 * float64(packagesActives[k]) / float64(packagesTotals[k])
+		if v > int64(math.Floor(actual*10+0.5)) {
+			log.Printf("package %s coverage is %.1f%%, required %.1f%%\n", k, actual, float64(v)/10.0)
```
Comment (wxiaoguang) [inline build/gocoverage.go]: No `\n` for "log.Printf`

File: build/gocoverage.go [hunk 8]
```diff
@@ -99,12 +102,52 @@ func dumpProfiles(profiles []*cover.Profile, out io.Writer) {
...
+
+Commands:
+  %[1]s merge ...
+  %[1]s check [coverage_file] [package=percent] ...
+
+Arguments:
+  {file-list}     the file list
+
+Example:
+  %[1]s merge -s -d {file-list}
```
Comment (wxiaoguang) [inline build/gocoverage.go]: What is `-s` / `-d`?

File: .github/workflows/pull-db-tests.yml (hunk 3 above)
Comment (silverwind) [reply .github/workflows/pull-db-tests.yml]: Yes, [e2e tests](https://github.com/go-gitea/gitea/blob/main/.github/workflows/pull-e2e-tests.yml) are distinct from backend tests and they do of course run with javascript inside a headless browser. Thought I have to mention that E2E tests are lacking, but that is the overall idea.

Comment (silverwind) [inline .github/workflows/pull-db-tests.yml]:
```suggestion
    if: needs.files-changed.outputs.backend == 'true'
```

Frontend coverage is a separate topic and not related.

File: Makefile [hunk 9]
```diff
@@ -220,6 +220,7 @@ help:
 	@echo " - test                             test everything"
 	@echo " - test-frontend                    test frontend files"
 	@echo " - test-backend                     test backend files"
+	@echo " - test-coverage-check              check package coverage in tests"
```
Comment (silverwind) [inline Makefile]: `check-backend-coverage`?

**Comment by silverwind** [comment]
Will there be an option to ignore reduced coverage? Like when some stuff is hard/impossible to test for, e.g. admin should still be able to merge.

File: build/gocoverage.go (hunk 5 above)
Comment (lunny) [reply build/gocoverage.go:169]: Some output will be `12.3%`, and I think it's accurate enough.

Comment (wxiaoguang) [reply build/gocoverage.go:169]: I would say this trick is difficult to understand, and I do not understand why it must be casted to `int64`

IMO keep the original float "12.3" is clear enough, or unify the unit to natural unit (1.0 mean 100%, 0.02 mean 2%)

File: .github/workflows/pull-db-tests.yml [hunk 10]
```diff
@@ -251,3 +263,17 @@ jobs:
...
+    needs: [files-changed, test-pgsql, test-sqlite, test-unit, test-mysql5, test-mysql8, test-mssql]
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - uses: actions/setup-go@v4
+      with:
+        go-version: ">=1.20"
+        check-latest: true
+    - uses: actions/download-artifact@v3
+    - run: go run build/gocoverage.go merge unit-test-coverage/coverage.out unit-test-coverage-gogit/coverage.out unit-test-coverage-integration/coverage.out > coverage.all
```
Comment (silverwind) [inline .github/workflows/pull-db-tests.yml:278]: Calls like this should be in the Makefile so that coverage can be produced locally as well. Please don't lock us in into actions with commands like this.

**Comment by silverwind** [comment]
Yes, I'm always asking where possible, I think it's the best we can do. Strict requirements will not work so good I guess.
